Update ezautolinkoperator.php

tompsonx opened the pull request:
Add support for text which is started from www. without protocol.
For example change text "www.ez.no" to link tags.

brookinsconsulting commented:
-1 without ini setting disabling this feature by default

tompsonx commented:
I added ini setting disabling this feature by default.

brookinsconsulting commented:
Hello @tompsonx

Thank you for you prompt response.

We have some additional comments but first would you create a jira.ez.no issue ticket and share the link here in the pr?

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,17 +42,48 @@ function formatUri( $url, $max )
         return "<a href=\"$url\" title=\"$url\">$text</a>";
     }
+    
+    static function formatUriSimple( $url, $max, $method = '' )
```
Do we really need a separate function? We really don't think so. We also do not care for the current name as vague. We think you features would be better placed within the original function instead.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -60,6 +91,7 @@ function modify( $tpl, $operatorName, $operatorParameters, $rootNamespace, $curr
     {
         $ini = $tpl->ini();
         $max = $ini->variable( 'AutoLinkOperator', 'MaxCharacters' );
+        $simpleWww = ( $ini->variable( 'AutoLinkOperator', 'ChangeUrlWithoutProtocol' ) == 'enabled' );
```
We think that this var name should be more clear and specific on what it represents. While perhaps too long even $linkWwwUrlWithoutProtocol or $linkWwwUrl would be better

brookinsconsulting commented on settings/template.ini:
```diff
@@ -72,6 +72,10 @@ Methods[]=ftps
 
+# Add support for text which is started from www. without protocol.
+# For example change text "www.ez.no" to link tags.
+ChangeUrlWithoutProtocol=disabled
```
We think that this setting name should be more clear and specific on what it represents. LinkWwwUrlsWithoutProtocol would be better because it is very clear and describes exactly what your feature does. Your current setting name is too vague (Re: use of terms Change and Url (singular)) and technically inaccurate as it only links urls starting with 'www.'.

tompsonx commented:
Thank you for your constructive comments.
I rebuilt a little code. What do you think about such modifications.
I added static properties, changed preg_replace to preg_replace_callback and added parameter for remove protocols from text.

I create jira issue: EZP-24114.

brookinsconsulting commented:
Link to issue ticket, https://jira.ez.no/browse/EZP-24114

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -10,6 +10,13 @@
 {
+    static $max = 72;
+    
+    static $isLinkWwwUrl = false;
```
This variable name is not so good, as it suggests by it's name (to us at any level) that the link in question is a www. style link but in truth the variable is only a boolean indicating that the operator use link www. type text. For these reasons we think this variable name should be changed. `$linkWwwUrls` (plural) is in our opinion a much better name. We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -10,6 +10,13 @@
+    
+    static $isLinkWwwUrl = false;
+    
+    static $hideProtoclText = '';
```
This variable name should also be changed to `$hideProtocolsText` at the very least. We think this was merely a spelling typo. We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -33,7 +40,12 @@ function namedParameterList()
+     Leave for backward compatibilty.
+     \static
+     \deprecated
+    */
```
For all class methods you touch / change / modify (at least) in this class file please convert the function doxygen comments to phpdoc (the code standard for all new eZ Publish Legacy code). For all new methods please create new phpdoc function comments. This step is required.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
         return "<a href=\"$url\" title=\"$url\">$text</a>";
     }
+    
+    static function formatUriCallback( $matches )
```
This method needs a phpdoc comment block above it

brookinsconsulting replied:
This class method name still leaves something to be desired. We do not think that 'Callback' method name text is clear or needed. We suggest `formatUrlLinkHtml` as a replacement for this class method name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+    static function formatUriCallback( $matches )
+    {
+        $method = $matches[1];
+        $www = $matches[2];
```
This variable name should also be changed to `$hostname` at the very least. We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+    {
+        $method = $matches[1];
+        $www = $matches[2];
+        $domainPath = $matches[3];
```
This variable name should also be changed at the very least. We suggest `$domainAndUri` as a more clear and verbose replacement. We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+            $url = 'http://' . $www . $domainPath;
+        }
+        
+        $textFull = $url;
```
We think this variable name should also be changed. We suggest `$urlLinkText` as a more clear and verbose replacement. We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+            $textFull = preg_replace( '`^(?:' . self::$hideProtoclText . '):\/\/(.*)$`', '$1', $textFull );
+        }
+        
+        $text = $textFull;
```
Please add a blank line here for increased readability.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+        if (strlen($text) > self::$max)
+        {
+            $text = substr($text, 0, (self::$max / 2) - 3). '...'. substr($text, strlen($text) - (self::$max / 2));
+        }
```
Please add a blank line here for increased readability.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
-        return preg_replace(
-            "`(?<!href=\"|href='|src=\"|src='|value=\"|value=')($methods):\/\/[\w]+(.[\w]+)([\w\-\.,@?^=%&:\/~\+#;*\(\)\!]*[\w\-\@?^=%&\/~\+#;*\(\)\!])?`e",
-            'eZAutoLinkOperator::formatUri("$0", '. $max. ')',
+        if ( $max !== false && is_int( $max ) )
```
Please use if conditional statement brackets here to conform to the coding standard and increase readability.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+        if ( $max !== false && is_int( $max ) )
+            self::$max = $max;
+        
+        if ( self::$isLinkWwwUrl )
```
Please use if conditional statement brackets here to conform to the coding standard and increase readability.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -42,15 +54,52 @@ function formatUri( $url, $max )
+            self::$max = $max;
+        
+        if ( self::$isLinkWwwUrl )
+            $mainSearchText = "(?:((?:$methods):\/\/)|(www\.))";
```
This variable name is not so good, it's simply just not very clear. We think that `$searchExpPart`is a (somewhat) more accurate name (at least). We request you change this variable name.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -59,20 +108,25 @@ function addURILinks( $text, $max, $methods = 'http|https|ftp' )
     {
         $ini = $tpl->ini();
-        $max = $ini->variable( 'AutoLinkOperator', 'MaxCharacters' );
+        self::$max = $ini->variable( 'AutoLinkOperator', 'MaxCharacters' );
```
Please add a blank line here for increased readability.

brookinsconsulting commented on kernel/common/ezautolinkoperator.php:
```diff
@@ -59,20 +108,25 @@ function addURILinks( $text, $max, $methods = 'http|https|ftp' )
 
+        self::$isLinkWwwUrl = ( $ini->variable( 'AutoLinkOperator', 'LinkWwwUrlsWithoutProtocol' ) == 'enabled' );
+        
+        $hideProtocls = $ini->variable( 'AutoLinkOperator', 'HideProtocols' );
```
This variable name should also be changed to `$stripProtocolsText` at the very least. We think this was merely a spelling typo. We request you change this variable name.

brookinsconsulting commented on settings/template.ini:
```diff
@@ -72,6 +72,13 @@ Methods[]=ftps
+# For example change text "www.ez.no" to link tags.
+LinkWwwUrlsWithoutProtocol=disabled
+
+HideProtocols[]
```
This variable name is not so good, as it suggests by it's name (to us at any level) that the protocol text is hidden, which is not exactly true. Instead in truth, when this feature is used the protocols are striped from the generated link text. For these reasons we think this setting name should be changed. `StripLinkProtocols` is in our opinion a better name. We request you change this setting name.

brookinsconsulting replied:
This setting block requires a detailed comment explaining it's use. Here is a suggested example:
```
# Strip protocols from link text when using autolink template operator
# For example link text string "http://www.ez.no". Disabled by default to retain BC.
# Add link protocols to enable this feature.```

brookinsconsulting commented on settings/template.ini:
```diff
@@ -72,6 +72,13 @@ Methods[]=ftps
+LinkWwwUrlsWithoutProtocol=disabled
+
+HideProtocols[]
+#HideProtocols[]=http
```
Now for a more positive comment :) We like that you give a specific example of how to use this setting.

brookinsconsulting commented on settings/template.ini:
```diff
@@ -72,6 +72,13 @@ Methods[]=ftps
 Methods[]=webdavs
 
+# Add support for text which is started from www. without protocol.
+# For example change text "www.ez.no" to link tags.
```
We think this setting comment block text should be re-written and clarified. Here is a suggested replacement:

`# Link text urls starting with 'www.' without protocol when using autolink template operator
# For example link text string "www.ez.no". Disabled by default to retain BC.
# Use either enabled or disabled.`

brookinsconsulting commented:
Hello @tompsonx 

Apologies for so many individual refactoring requests / comments.

We are striving to ensure that if this pull request is merged, it is done only when the code is of the very best possible quality.

In eZ Publish Legacy we are striving to ensure BC, strict code standards and reliability. In general these days we are actually trying very hard to *not* make large or unnecessary changes to the code.

For these reasons and (specifically) because of your regex changes. I think for this to be considered for merging we would prefer you also provide extensive unit test coverage of this class covering all previous / existing and new use cases.

Functions that would need to be tested would be those you added and modified. The goal is to make sure they behave the way they are supposed to in each unique use case (variable input use cases, and by our mark, there are quite a few which will need to be written).

Unit test are using phpunit (you can start by looking at tutorials online). You can find documentation on how to run legacy unit test here: http://www.ezpedia.org/ez/testing_ez_publish_test_system

You can add new tests in: https://github.com/ezsystems/ezpublish-legacy/blob/master/tests/tests/kernel/common/ezautolinkoperator_test.php

Cheers,
Brookins Consulting
